### ISSUE TITLE
index.d.ts needs DOM in lib to work

### ISSUE THREAD

**Issue by lantw44** [Automattic/mongoose#10504]
**Do you want to request a *feature* or report a *bug*?**
Bug

**What is the current behavior?**
It doesn't compile if the tsconfig.json doesn't include `DOM` in `lib` option. I am not going to run Mongoose on a browser, so I should not be required to include DOM libraries in the environment.
```
node_modules/mongoose/index.d.ts:191:9 - error TS2304: Cannot find name 'WritableStream'.

191       | WritableStream
            ~~~~~~~~~~~~~~


Found 1 error.
```
`git blame` said `WritableStream` came from https://github.com/Automattic/mongoose/pull/10471.

**If the current behavior is a bug, please provide the steps to reproduce.**
Compile this project.

package.json
```json
{
  "name": "test",
  "version": "1.0.0",
  "license": "MIT",
  "dependencies": {
    "mongoose": "^5.13.4"
  },
  "devDependencies": {
    "@types/node": "^14.17.6",
    "typescript": "~4.3.5"
  }
}
```

tsconfig.json
```json
{
  "compilerOptions": {
    "target": "ES2019",
    "module": "commonjs",
    "lib": ["ES2019"],
    "strict": true,
    "esModuleInterop": true,
    "skipLibCheck": false,
    "forceConsistentCasingInFileNames": true
  }
}
```

index.ts
```
import mongoose from 'mongoose';
```

**What is the expected behavior?**
It should compile without DOM.

**What are the versions of Node.js, Mongoose and MongoDB you are using? Note that "latest" is not a version.**
Node.js 14.17.0
Mongoose 5.13.4
MongoDB 4.4.7

**Comment by thiagokisaki** [comment]
@vkarpov15 

It looks like there are two different interfaces with the same name:
<img src="https://user-images.githubusercontent.com/66492727/127527508-b45a2baa-d1ec-4340-af46-c254b7b378c5.png" width="500" />
<img src="https://user-images.githubusercontent.com/66492727/127527593-6cc41e26-29b1-4c7c-882f-0e3e84e226b9.png" width="500" />

I should have used `NodeJS.WritableStream` or, more strictly, `stream.Writable`, because of the following check in `/lib/drivers/node-mongodb-native/collection.js`:
![Screenshot from 2021-07-29 13-10-00](https://user-images.githubusercontent.com/66492727/127529143-0f344300-c3b6-4612-8e69-c46418d64ac4.png)

**Comment by sigkar1** [comment]
I have this issue too. Version 5.13.4 is broken I get the same error. Please advise.
the error

node_modules/mongoose/index.d.ts:191:9 - error TS2304: Cannot find name 'WritableStream'.

191       | WritableStream

**Comment by hasezoey** [comment]
I have this issue too, noticed it while trying out 6.0, project's used `@types/node` version is `~12.12.6` (project's lowest supported nodejs version) (not having tsconfig `compilerOptions.lib.dom` enabled)

it can currently (if wanted) easily reproduced by cloning [this repo](https://github.com/typegoose/typegoose) and checking out the branch `mongoose6.0` and then running `yarn install` and `yarn run build` and the first tsc error that comes up should be about this issue

or an easier light-weight repro:
- clone [this repo](https://github.com/typegoose/typegoose-testing)
- checkout `reproMongoose10504`
- run `yarn install`
- run `yarn run run` (even without having a mongodb instance up)

full tsc error:
```
node_modules/mongoose/index.d.ts:191:9 - error TS2304: Cannot find name 'WritableStream'.

191       | WritableStream
            ~~~~~~~~~~~~~~


Found 1 error.
```

**Comment by lantw44** [comment]
> I'm unable to repro this, can you please confirm what version of `@types/node` you have installed? 

Yes.
```
$ find node_modules -name node
node_modules/mongoose/node_modules/@types/node
node_modules/@types/node
node_modules/@types/mongodb/node_modules/@types/node
node_modules/@types/bson/node_modules/@types/node
```
package.json has `"@types/node": "^14.17.6"`, and it matches the installed version.
```
$ jq .version < node_modules/@types/node/package.json 
"14.17.6"
$ jq .version < node_modules/mongoose/node_modules/@types/node/package.json 
"15.14.4"
$ jq .version < node_modules/@types/mongodb/node_modules/@types/node/package.json 
"16.4.7"
$ jq .version < node_modules/@types/bson/node_modules/@types/node/package.json 
"16.4.7"
```
Reproduce the bug with the files I provided in the description:
```
$ node --version
v14.17.0
$ npm --version
6.14.13
$ ls -l
total 12
-rw-r--r--. 1 lantw44 lantw44  33 Jul 30 11:15 index.ts
-rw-r--r--. 1 lantw44 lantw44 200 Jul 30 11:14 package.json
-rw-r--r--. 1 lantw44 lantw44 225 Jul 30 11:15 tsconfig.json
$ npm install
npm notice created a lockfile as package-lock.json. You should commit this file.
npm WARN test@1.0.0 No description
npm WARN test@1.0.0 No repository field.

added 38 packages from 96 contributors and audited 38 packages in 1.491s

2 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities

$ npx tsc
node_modules/mongoose/index.d.ts:191:9 - error TS2304: Cannot find name 'WritableStream'.

191       | WritableStream
            ~~~~~~~~~~~~~~


Found 1 error.

(npx tsc exited with status 2)
```

**Comment by vkarpov15** [comment]
Oddly enough I'm still unable to repro this, even with `@types/node@12`. Maybe it's a yarn issue? Either way, not worth debating too much, #10510 works for me as well.

**Comment by acomito** [comment]
having the same issue with

```
    "mongoose": "^5.13.4",
```

```
{
    "compilerOptions": {
      "allowJs": true,
      "module": "commonjs",
      "moduleResolution": "node",
      "pretty": true,
      "sourceMap": true,
      "target": "es6",
      "outDir": "./dist",
      "lib": ["es6"],
      "resolveJsonModule": true,
      "types": ["node", "graphql", "express"],
      "typeRoots": ["./node_modules/@types", "./src/@types"],
      "experimentalDecorators": true,
      "emitDecoratorMetadata": true,
      "esModuleInterop": true,
    },
    "include": ["src/entities/**/*.ts", "src/**/*.ts", "src/**/*.js", "./index.ts"],
    "exclude": ["node_modules","ormlogs.log", "dist"]
  }
```

**Comment by hasezoey** [comment]
@acomito for me (and probably others), this issue was fixed by version `5.13.5`, try to upgrade to that version (from `5.13.4`)

**Comment by vkarpov15** [comment]
@acomito if @hasezoey 's suggestion doesn't work, please open a new issue and follow the issue template.
